accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +577,282 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
+    externally provided initializer list and/or body.*
 
-A typedef can't be used to subvert these restrictions. When extending,
-implementing, or mixing in a typedef, we look at the library where class or
-mixin the typedef resolves to is defined to determine if the behavior is
-allowed. *Note that the library where the _typedef_ is defined does not come
-into play. Typedefs cannot be marked with any of the new modifiers.*
+A trivial constructor may be named or unnamed, and `const` or non-`const`. A
+*non-trivial generative constructor* is a generative constructor which is not a
+trivial generative constructor. Examples:
+
+```dart
+class C {
+  // Trivial generative constructors:
+  C();
+  const C();
+
+  // Non-trivial generative constructors:
+  C(int x);
+  C(this.x);
+  C() { ... }
+  C(): super();
+
+  // Not generative constructors, so neither trivial generative nor non-trivial
+  // generative:
+  factory C.f = C;
+  factory C.f2() { ... }
+}
+```
+
+It's a compile-time error if:
+
+*   A `mixin class` declaration has a superclass other than `Object`. *The
+    declaration is limited to an `extends Object` clause or no `extends` clause,
+    and no `with` clauses. The class grammar prohibits `on` clauses.*
+
+*   A `mixin class` declaration declares any non-trivial generative constructor.
+    *It may declare no constructors, in which case it gets a default
+    constructor, or it can declare factory constructors and/or trivial
+    generative constructors.*
+
+These rules ensure that when you mark a `class` with `mixin` that it *can* be
+used as one.
+
+A class not marked `mixin` can still be used as a mixins when the class's
+declaration is in a pre-feature library. Post-feature libraries can also mix in
+their own classes that aren't marked `mixin` as long as the class supports it.
+Specifically:
+
+It's a compile-time error to for a declaration in library `L` to mix in a
+non-`mixin` class declaration `D` from library `K` if any of:
+
+*   The superclass of `D` is not `Object`,
+
+*   `K` is a pre-feature library, and `D` declares any constructors, or
+
+    *For pre-feature libraries, we can't tell if the intent of `class` was "just
+    a class" or "both a class and a mixin". For compatibility, we assume the
+    latter, even if the class is being used as a mixin in a post-feature library
+    and where it does happen to be possible to distinguish those two intents.*
+
+*   `K` is a post-feature library, and any of:
+
+    *   `K` is not the same library as `L`,
+
+    *   `D` is a mixin-application class declaration (has the form `class D =
+        ...;`), or
+
+    *   `D` declares any non-trivial generative constructors.
+
+    *When a class is in a library where it possible to distinguish between
+    whether the class is intended to use it as a mixin or not, the author is
+    obligated to document the intent, and that intent applies to all other
+    libraries regardless of their version.*
```

Comment:
Maybe worth saying that A class _can_ mix in a non-mixin class from the same library, if it has `Object` as supertype and no non-trivial generative constructors. Adding `mixin` to a class declaration only changes whether *other* libraries can do.

(And as always, that's because you only risk breaking your own library when changing the class, and you should know immediately and be in a position to fix it.)

Reply:
I did not add that, based on the expectation that this will in fact be an error.